include/rocksdb/write_buffer_manager.h

```diff
@@ -14,48 +14,78 @@
 
 #pragma once
 
+#include <stdio.h>
 #include <atomic>
 #include <cstddef>
+#include "rocksdb/cache.h"
 
 namespace rocksdb {
 
 class WriteBufferManager {
  public:
-  // _buffer_size = 0 indicates no limit. Memory won't be tracked,
+  // _buffer_size = 0 indicates no limit. Memory won't be capped.
   // memory_usage() won't be valid and ShouldFlush() will always return true.
-  explicit WriteBufferManager(size_t _buffer_size)
-      : buffer_size_(_buffer_size), memory_used_(0) {}
-
-  ~WriteBufferManager() {}
+  // if `cache` is provided, we'll put dummy entries in the cache and cost
+  // the memory allocated to the cache. It can be used even if _buffer_size = 0.
+  explicit WriteBufferManager(size_t _buffer_size,
+                              std::shared_ptr<Cache> cache = {});
+  ~WriteBufferManager();
 
   bool enabled() const { return buffer_size_ != 0; }
 
   // Only valid if enabled()
   size_t memory_usage() const {
     return memory_used_.load(std::memory_order_relaxed);
   }
+  size_t mutable_memtable_memory_usage() const {
+    return mutable_memtable_used_.load(std::memory_order_relaxed);
+  }
   size_t buffer_size() const { return buffer_size_; }
 
   // Should only be called from write thread
   bool ShouldFlush() const {
-    return enabled() && memory_usage() >= buffer_size();
+    // printf("%ld %ld\n", (long) memory_usage(), (long)
+    // mutable_memtable_memory_usage());
+    // Flash if memory usage hits a hard limit, or total mutable memtable size
+    // hits a soft limit, which is 7/8 of the hard limit.
+    return enabled() &&
+           (memory_usage() >= buffer_size() ||
+            mutable_memtable_memory_usage() >= buffer_size() * 7 / 8);
   }
 
-  // Should only be called from write thread
   void ReserveMem(size_t mem) {
-    if (enabled()) {
+    if (cache_rep_ != nullptr) {
+      ReserveMemWithCache(mem);
+    } else if (enabled()) {
       memory_used_.fetch_add(mem, std::memory_order_relaxed);
     }
+    if (enabled()) {
+      mutable_memtable_used_.fetch_add(mem, std::memory_order_relaxed);
+    }
   }
-  void FreeMem(size_t mem) {
+  void ImmuteMemtable(size_t mem) {
```

Comment:
Because it is called when we make a memtable immutable. How about call it ScheduleFreeMem()?